metadata-ingestion/src/datahub/ingestion/source/glue.py

```diff
@@ -131,44 +151,270 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
+
+        paginator = self.glue_client.get_paginator("get_jobs")
+        for page in paginator.paginate():
+            jobs += page["Jobs"]
+
+        return jobs
+
+    def get_dataflow_graph(self, script_path: str) -> Dict[str, Any]:
+        """
+        Get the DAG of transforms and data sources/sinks for a job.
+
+        Parameters
+        ----------
+            script_path:
+                S3 path to the job's Python script.
+        """
+
+        # extract the script's bucket and key
+        url = urlparse(script_path, allow_fragments=False)
+        bucket = url.netloc
+        key = url.path[1:]
+
+        # download the script contents
+        obj = self.s3_client.get_object(Bucket=bucket, Key=key)
+        script = obj["Body"].read().decode("utf-8")
+
+        # extract the job DAG from the script
+        return self.glue_client.get_dataflow_graph(PythonScript=script)
+
+    def process_dataflow_graph(self, dataflow_graph, flow_urn):
+        """
+        Prepare a job's DAG for ingestion.
+
+        Parameters
+        ----------
+            dataflow_graph:
+                Job DAG returned from get_dataflow_graph()
+            flow_urn:
+                URN of the flow (i.e. the AWS Glue job itself).
+        """
+
+        new_dataset_ids = []
+        new_dataset_mces = []
+
+        def process_node(node):
+            """
+            Process a single node in the DAG.
+            """
+
+            node_type = node["NodeType"]
+
+            # for nodes representing datasets, we construct a dataset URN accordingly
+            if node_type in ["DataSource", "DataSink"]:
+
+                node_args = {x["Name"]: json.loads(x["Value"]) for x in node["Args"]}
+
+                # if data object is Glue table
+                if "database" in node_args and "table_name" in node_args:
+
+                    full_table_name = (
+                        f"{node_args['database']}.{node_args['table_name']}"
+                    )
+
+                    # we know that the table will already be covered when ingesting Glue tables
+                    node_urn = f"urn:li:dataset:(urn:li:dataPlatform:glue,{full_table_name},{self.env})"
+
+                # if data object is S3 bucket
+                elif node_args.get("connection_type") == "s3":
+
+                    # remove S3 prefix
+                    s3_name = node_args["connection_options"]["path"][5:]
+
+                    if s3_name.endswith("/"):
+                        s3_name = s3_name[:-1]
+
+                    node_urn = (
+                        f"urn:li:dataset:(urn:li:dataPlatform:s3,{s3_name},{self.env})"
+                    )
+
+                    dataset_snapshot = DatasetSnapshot(
+                        urn=node_urn,
+                        aspects=[],
+                    )
+
+                    dataset_snapshot.aspects.append(Status(removed=False))
+                    dataset_snapshot.aspects.append(
+                        OwnershipClass(
```

Comment:
Would be great if there were a way to resolve owners from the AWS IAM roles associated with this table.. 

Or the schema of the S3 file. I guess we'd need something to parse the extension / file to do that?

Reply:
Yeah so this is already done for owners – it will take another query to get it for S3 files but I think it is possible.